### PR TITLE
Add `RETURNING` support for MariaDB 10.5.1+

### LINKED ISSUE #177: Support `RETURNING` for MariaDB
MariaDB 10.5.1+ support to use `INSERT ... RETURNING col1[, col2[, col3...]]` for returns a result-set of the inserted rows. The response looks like the execution result of a `SELECT` statement, and the last `OkMessage` will NOT contain the updated rows.

It will change the logic of `MySqlStatement.returnGeneratedValues` in MariaDB 10.5.1+ .

See also https://mariadb.com/kb/en/insertreturning/

### PR DESCRIPTION
**Pull request by mirromutth**
Motivation:

Add `RETURNING` clause support for MariaDB 10.5.1+ .

See also #177 .

Modification:

Add `RETURNING` clause in `MySqlStatement#returnGeneratedValues` and query message packets.

Result:

For MariaDB 10.5.1 and above, `MySqlStatement#returnGeneratedValues` will use `RETURNING` clause instead of `LAST_INSERT_ID`.

Notice:

For MariaDB 10.5.0 and below, `MySqlStatement#returnGeneratedValues` will still use `LAST_INSERT_ID` due to these versions do not support `RETURNING` clause.

### CODE REVIEW

**Comment by mirromutth** [comment]
@jchrys

Currently, it is simplify to add `RETURNING` following original SQL statement without checking.

Perhaps the statement parser needs to be rewritten to support checking whether the statement supports `RETURNING` (`INSERT`/`REPLACE`/`DELETE`) or already includes `RETURNING`, like [`r2dbc-mariadb` does](https://github.com/mariadb-corporation/mariadb-connector-r2dbc/blob/master/src/main/java/org/mariadb/r2dbc/MariadbClientParameterizedQueryStatement.java#L126)?

**Comment by jchrys** [comment]
> @jchrys
> 
> Currently, it is simplify to add `RETURNING` following original SQL statement without checking.
> 
> Perhaps the statement parser needs to be rewritten to support checking whether the statement supports `RETURNING` (`INSERT`/`REPLACE`/`DELETE`) or already includes `RETURNING`, like [`r2dbc-mariadb` does](https://github.com/mariadb-corporation/mariadb-connector-r2dbc/blob/master/src/main/java/org/mariadb/r2dbc/MariadbClientParameterizedQueryStatement.java#L126)?

I think  the current change is suitable for merging, as MariaDB will naturally return a syntax error in case of incorrect `RETURNING` usage. I agree that enhancing the parser, akin to what's done in `r2dbc-mariadb`, better be next step.

File: src/main/java/io/asyncer/r2dbc/mysql/MySqlResult.java [hunk 1]
```diff
@@ -309,10 +315,17 @@ public void accept(ServerMessage message, SynchronousSink<Segment> sink) {
 
                 this.rowMetadata = MySqlRowMetadata.create(metadataMessages);
             } else if (message instanceof OkMessage) {
-                Segment segment = generatedKeyName == null ? new MySqlUpdateCount((OkMessage) message) :
-                    new MySqlOkSegment((OkMessage) message, codecs, generatedKeyName);
+                OkMessage msg = (OkMessage) message;
+
+                if (MySqlStatementSupport.supportReturning(context) && msg.isEndOfRows()) {
+                    sink.next(new MySqlUpdateCount(rowCount.get()));
```
Comment (jchrys) [inline src/main/java/io/asyncer/r2dbc/mysql/MySqlResult.java]: Do might need to add `rowCount.set(0)` after this line?

**Review by jchrys — APPROVED** [review]
lgtm 👍

**Comment by mirromutth** [comment]
Columns should not be quoted, it can be an expression. e.g. `INSERT INTO ... RETURNING CURRENT_TIMESTAMP`.

If we quoted it, like in the example above it would be <code>INSERT INTO ... RETURNING &#96;CURRENT_TIMESTAMP&#96;</code>, which would be confusing. User wants the result of the `CURRENT_TIMESTAMP` expression, not a column named `CURRENT_TIMESTAMP`.

Keep columns raw and let ORM/QueryBuilder handle it.

**Comment by mirromutth** [comment]
@jchrys It is foreseeable that this PR will change the behavior of `returnGenerateValues` when connecting to MariaDB 10.5.1+, maybe we should merge it with the next minor version?

For example, if a user is using `returnGenerateValues` without arguments, we will return a `Result` with one and only one row, which contains a `LAST_INSERT_ID` column. This modification will change the number of result rows (maybe more, maybe 0 by `INSERT ... SELECT ... RETURNING ...`), and the row metadata will contain all data columns in the table.

**Comment by jchrys** [comment]
> @jchrys It is foreseeable that this PR will change the behavior of `returnGenerateValues` when connecting to MariaDB 10.5.1+, maybe we should merge it with the next minor version?
> 
> For example, if a user is using `returnGenerateValues` without arguments, we will return a `Result` with one and only one row, which contains a `LAST_INSERT_ID` column. This modification will change the number of result rows (maybe more, maybe 0 by `INSERT ... SELECT ... RETURNING ...`), and the row metadata will contain all data columns in the table.

I agree. 😄

**Review by jchrys — APPROVED** [review]

**Pull request merged**